### DISCUSSION TITLE
Is it possible to use this library with VS Code?

### DISCUSSION THREAD

**Discussion by izamoramx** [JAndrassy/ArduinoOTA#59] [General]
Hello @jandrassy ,

Is there any documentation on how to configure VS Code projects to be able to upload via IP address with your library? Or could you share in here how to configure the VS Code to make it work.

I have everything working on Arduino IDE but I would like to be able to do the same on VS Code. (all the code compiles, just fails during upload).

Thanks,

**Comment by JAndrassy** [comment]
Platformio?

**Comment by izamoramx** [comment]
> Platformio?

No. Just VS Code with the MS Arduino for Visual Studio Code plug in.

**Comment by JAndrassy** [comment]
you can use the library, but I know nothing about the VS Arduino plugin. does it use Arduino boards packages? will it read platform.local.txt for additional build and upload commands? does it show the network port?
if not, you can do upload from command line. the upload command is in platform.local.txt in extras folder. .bin file must be uploaded so for AVR the hex file must first be converted to .bin

**Comment by izamoramx** [comment]
I haven't been able to find anything specific saying that it does or doesn't. As far as I am aware, from the available information, it is just an IDE for Arduino. I requires arduino IDE to be installed and it uses it to compile. It uses all the preinstall libraries. Other than the specific features of VS Code, I don't see it using it's own libraries or code to generate the binaries so I believe it just interfaces with Arduino IDE. Here is the link to the project FYI:

https://marketplace.visualstudio.com/items?itemName=vsciot-vscode.vscode-arduino

I believe it works the same as Eclipse, and I think I saw you mentioned on a conversation that you use Eclipse to develop for Arduino. Is there any configuration required to make OTA work on Eclipse? maybe the same thing could solve the issue on VS Code.

**Comment by JAndrassy** [comment]
I use Eclipse Sloeber plugin and that doesn't require Arduino IDE installation but uses Arduino boards packages (has own way to install them). But Sloeber doesn't read platform.local.txt patch so the changes must be merged to platform.txt or as I use it with the [my_boards](https://github.com/jandrassy/my_boards) concept.

**Comment by anklimov** [comment]
It's absolutely possible to use lib with platformio+vsc, target config example from platformio.ini below:
Note: be sure you have arduinoOTA utility somewhere in PATH

[env:esp32-wifi]
platform = espressif32
framework = arduino
monitor_baud = 115200
board = esp32-evb
upload_command = arduinoOTA -address 192.168.1.1 -port 65280 -username arduino -password password -b -upload /sketch -sketch $SOURCE
upload_protocol = custom

**Reply by entozoon** [reply]
That's awesome, but how do you install the arduinoOTA utility? Is it part of the Arduino IDE and ArdiunoOTA library therein? Didn't really want to have to install all that..

**Reply by anklimov** [reply]
Initially, ArduinoOTA utility is part of Arduino IDE. But, in cases of big binaries, original version is non usable due short timeout. You can change timeout to some appropriate value, re-compile it, using GO and put executable file somewhere in PATH 

I've done it already - https://github.com/anklimov/arduinoOTA-client
(included pre-compiled binaries for all platforms)

Another reasonable alternative - use curl utility instead

**Reply by entozoon** [reply]
That worked perfectly, thank you so much @anklimov! I can finally programme OTA in VSC!

**Reply by izamoramx** [reply]
@anklimov I was able to implement this solution after troubleshooting various issues along the way, but I think I am very close to making the OTA fully work. Currently I am able to do it trough the Arduino IDE with no problem multiple times in a row, but when I try over VS Code and PlatformIO I always get this error:

Configuring upload protocol...
AVAILABLE: atmel-ice, blackmagic, custom, jlink, sam-ba
CURRENT: upload_protocol = custom
Uploading .pio\build\mkrwifi1010\firmware.bin
Connecting to board ...  done
Uploading sketch ...  done
Flashing sketch ...  done
Error flashing the sketch:Unauthorized
*** [upload] Error 1
=============================================================================== [FAILED] Took 31.62 seconds

I have made sure that the password given is correct. I even copied the same upload commands as Arduino IDE uses (varies a little from the one you provided) on both your version of the .exe file and the one used by the IDE, and it always gives me the same error.

I have confirmed that the sketch was not uploaded even though is said "Flashing sketch...done"

Any idea what the issue might be? Your help will be greatly appreciated.

[UPDATE]: After a lot of research and tinkering with various things it seems that the culprit was the MS Windows Defender Firewall. I tried turning it off and uploading the code and it worked without an issue. I thought I was going to have to update the firewall tables but after turning it back on and running the upload again, it worked without needing any changes. Hopefully this solution sticks. Thank you for the valuable answers on this discussion. I hope this info helps others with the same issue.

[UPDATE 01/19/2022]: It seems that the previous solution of turning the firewall off was short-lived. I am back at getting the same error from before. @entozoon or @anklimov  I would really appreciate some help here as I am currently stuck! :(

**Comment by izamoramx** [comment]
> It's absolutely possible to use lib with platformio+vsc, target config example from platformio.ini below:
> Note: be sure you have arduinoOTA utility somewhere in PATH
> 
> [env:esp32-wifi]
> platform = espressif32
> framework = arduino
> monitor_baud = 115200
> board = esp32-evb
> upload_command = arduinoOTA -address 192.168.1.1 -port 65280 -username arduino -password password -b -upload /sketch -sketch $SOURCE
> upload_protocol = custom

@anklimov  thanks for the reply. Do you know if there is a similar configuration for just using the VS Code Arduino plugin instead of using PlatformIO?

**Reply by anklimov** [reply]
Unfortunatelly, I don't tried VS Code Arduino plugin, just Platformio. Hope, it's possible to configure something similar using command line above
